src/Lorentz/Contracts/GameDAO.hs

```diff
@@ -0,0 +1,203 @@
+  annOptions = DAO.baseDaoAnnOptions
+
+type NewContents = [NewContent]
+
+data NewContent = NewContent
+  { ncContentName :: MText
+  , ncContentDescription :: [MText]
+  }
+  deriving stock (Generic)
+  deriving anyclass (IsoValue)
+
+instance TypeHasDoc NewContent where
+  typeDocMdDescription = "Describe a new content. The content should have a \
+    \name and detail description."
+
+instance HasAnnotation NewContent where
+  annOptions = DAO.baseDaoAnnOptions
+
+instance TypeHasDoc GameDaoProposalMetadata where
+  typeDocMdDescription = "GameDAO's metadata. This fields affect how proposal \
+    \got accepted and how many tokens will be frozen"
+
+----------------------------------------------------------------------------
+
+type instance ErrorArg "fAIL_DECISION_LAMBDA" = ()
+
+instance CustomErrorHasDoc "fAIL_DECISION_LAMBDA" where
+  customErrClass = ErrClassActionException
+  customErrDocMdCause = "Trying to execute decision lambda but result in errors."
+
+----------------------------------------------------------------------------
+
+type Parameter = DAO.Parameter GameDaoProposalMetadata
+type Storage = DAO.Storage GameDaoProposalMetadata
+
+emptyStorage :: Storage
+emptyStorage = DAO.emptyStorage
+  { DAO.sVotingPeriod = 60 * 60 * 24 * 7 -- 7 days
+  , DAO.sQuorumThreshold = 100 -- All player votes which show that this proposals count
+  }
+
+-- | For "balance change", the proposer needs to freeze 10 tokens
+-- For "new content", the proposal needs to freeze 50 tokens
+gameDaoProposalCheck
+  :: forall s. (DAO.ProposeParams GameDaoProposalMetadata) : s
+  :-> Bool : s
+gameDaoProposalCheck = do
+  getFieldNamed #ppFrozenToken; swap
+  toField #ppProposalMetadata
+  toField #pmProposalType
+  stackType @(ProposalType : "ppFrozenToken" :! Natural : s)
+  caseT
+    ( #cBalanceType /-> do drop; push (10 :: Natural)
+    , #cNewType /-> do drop; push (50 :: Natural)
+    )
+  toNamed #requireValue
+
+  if #requireValue <=. #ppFrozenToken then
+    push True
+  else
+    push False
+
+-- | In case of a proposal got rejected, the proposer loses half of the tokens.
+gameDaoRejectedProposalReturnValue
+  :: forall s. (DAO.Proposal GameDaoProposalMetadata) : s
+  :-> ("slash_amount" :! Natural) : s
+gameDaoRejectedProposalReturnValue = do
+  toField #pProposerFrozenToken
+  push (2 :: Natural)
+  swap
+  ediv
+  ifSome car $
+    push (0 :: Natural)
+  toNamed #slash_amount
+
+-- | When a proposal got accepted, call a consumer contract with the proposal
+-- description. It would be more useful to send the whole proposal itself but
+-- using just the description is simpler to test.
+decisionLambda
+  ::  forall s store. (DAO.StorageC store GameDaoProposalMetadata)
+  =>  (DAO.Proposal GameDaoProposalMetadata) : store : s
+  :-> (List Operation, store) : s
+decisionLambda = do
+  toField #pMetadata
+  getField #pmProposalDescription
+
+  dip $ do
+    toField #pmConsumerAddr
+    contractCallingUnsafe @MText DefEpName
```

Comment:
Minor: AFAIU you can have `TAddress MText` in `pmConsumerAddr`, and then use plain `contractCalling CallDef` here.

Reply:
I tried to make this change, but it requires to add instance for `TypeHasDoc (TAddress MText)`. When I added it, this errors occur:

```
/home/rinne/project/serokell/baseDAO/src/Lorentz/Contracts/GameDAO.hs:110:10: error:
    • Type is not homomorphic: TAddress MText
    • In the expression:
        Michelson.Typed.Haskell.Doc.$dmtypeDocMdReference @(TAddress MText)
      In an equation for ‘typeDocMdReference’:
          typeDocMdReference
            = Michelson.Typed.Haskell.Doc.$dmtypeDocMdReference
                @(TAddress MText)
      In the instance declaration for ‘TypeHasDoc (TAddress MText)’
    |       
110 | instance TypeHasDoc (TAddress MText) where
```

Reply:
Oh, `TAddress` does not have this instance :fffuuu.

>     • In the expression:
>       Michelson.Typed.Haskell.Doc.$dmtypeDocMdReference @(TAddress MText)

`$dmtypeDocMdReference` means default implementation of `typeDocMdReference` so you need to initialize it with some other method. But ok, I think it does not worth doing right now.